utils/log.go

```diff
@@ -0,0 +1,96 @@
+package utils
+
+import "fmt"
+
+func FormatWritesLog(codes []byte, line *string) {
+	n := len(codes)
+	str := ""
+	exist := false
+	if n == 3 {
+		if str, exist = ASCIIGroupToStr(fmt.Sprintf("%X", codes)); exist {
+			*line += str
+			codes = nil
+		}
+	}
+	// sh control codes
+	if n >= 6 {
+		if str, exist = ASCIIGroupToStr(fmt.Sprintf("%X", []byte{codes[0], codes[1], codes[n-1]})); exist {
+			*line += str
+			codes = nil
+		}
+	}
+
+	if codes != nil {
+		str = ASCIIToStr(codes)
+		*line += str
+	}
+
+	return
+}
+
+func ASCIIToStr(codes []byte) string {
+	control := map[byte]string{
+		0:   "NUL",
+		1:   "SOH",
+		2:   "STX",
+		3:   "ETX",
+		4:   "EOT",
+		5:   "ENQ",
+		6:   "ACK",
+		7:   "BEL",
+		8:   "BS",
+		9:   "HT",
+		10:  "LF",
+		11:  "VT",
+		12:  "FF",
+		13:  "CR",
+		14:  "SO",
+		15:  "SI",
+		16:  "DLE",
+		17:  "DCI",
+		18:  "DC2",
+		19:  "DC3",
+		20:  "DC4",
+		21:  "NAK",
+		22:  "SYN",
+		23:  "TB",
+		24:  "CAN",
+		25:  "EM",
+		26:  "SUB",
+		27:  "ESC",
+		28:  "FS",
+		29:  "GS",
+		30:  "RS",
+		31:  "US",
+		32:  "SPACE",
+		127: "DEL",
+	}
+
+	str := ""
+	for _, code := range codes {
+		if value, ok := control[code]; ok {
+			str += value
+		} else {
+			str += string(code)
+		}
+	}
```

Comment:
As far as I can tell, this leaves us no way to tell the difference between hitting return or someone pasting the 2-character string "CR". That feels problematic.

Reply:
yes, in `utils/log.go` file, i can set `13: ""` to hide the `CR` control char, but i didn't. In order to clearly understand the user‘s behavior, I have displayed all the control characters.
Of course, it has some disadvantages, and I haven't figured out how to avoid them yet. At least it can handle most situations.

Reply:
I'm not suggesting you don't show the "CR". For auditing purposes, you need to know if the user hit Return or if they wrote "CR". Right now, there's no way to tell the difference. Maybe wrap the string from CtrlChar in brackets, so the output will read `[CR]`.  Any `[` typed by the user could be shown as `\[`. Any \`\ typed by the user could be shown as `\\`. Would that work?

Reply:
Great advice, I'll do it this way.